compiler/sem/pragmas.nim

```diff
@@ -1657,65 +1418,324 @@ proc prepareSinglePragma(
         sym.magic = mBoolDefine
       of wUsed:
         result = noVal(c, it)
-        if sym == nil:
-          result = invalidPragma(c, it)
-        else:
-          sym.flags.incl sfUsed
+        sym.flags.incl sfUsed
       of wEnforceNoRaises:
+        result = noVal(c, it)
         sym.flags.incl sfNeverRaises
+      of wCast:
+        result = c.config.newError(it, PAstDiag(kind: adSemCastRequiresStatement))
+      of wInvalid:
+        # a non-builtin pragma reaching here must be a custom pragma
+        result = it
       else:
-        result = invalidPragma(c, it)
-    elif comesFromPush and whichKeyword(ident) != wInvalid:
-      discard "ignore the .push pragma; it doesn't apply"
+        unreachable(k)
+
+proc processBlockPragma(c: PContext, it: PNode, p: TSpecialWord): PNode =
+  ## Pre-processes the pragma expression `it`, returning either the result or
+  ## an error. `p` is the name of the built-in pragma.
+  ##
+  ## Application of the pragmas happens later.
+  case p
+  of wLine:
+    pragmaLine(c, it)
+  of wLocks:
+    pragmaLockStmt(c, it)
+  of wNoRewrite, wGcSafe, wNoSideEffect:
+    noVal(c, it)
+  of wCast:
+    if whichPragma(it[1]) in {wTags, wRaises}:
+      let e = pragmaRaisesOrTags(c, it[1])
+      if e.kind == nkError:
+        e
+      else:
+        it
     else:
-      if sym == nil or (sym.kind in {skVar, skLet, skParam,
-                        skField, skProc, skFunc, skConverter, skMethod, skType}):
-        n[i] = semCustomPragma(c, it)
-        result = n[i]
-      elif sym != nil:
-        result = illegalCustomPragma(c, it, sym)
+      it
+  of wInvalid:
+    # must be a custom pragma -- they're allowed
+    it
+  of {low(TSpecialWord)..high(TSpecialWord)} - exprPragmas - {wCast, wInvalid}:
+    invalidPragma(c, it)
+
+proc applyStmtPragma(c: PContext, owner: PSym, it: PNode, k: TSpecialWord): PNode =
+  ## Evaluates the single built-in pragma statement `it` for its side-effects.
+  ## Depending on the pragma, this might mutate `owner`. If the statement
+  ## is ill-formed (e.g. missing arguments) or the evaluation failed, an error
+  ## is returned.
+  case k
+  of wUsed:
+    result = noVal(c, it)
+    # XXX: an escaping mutation if the the pragma is used inside a ``compiles``
+    #      block
+    owner.flags.incl sfUsed
+  of wDeprecated:
+    result = it
+    # TODO: disallow when the owner is not a module
```

Comment:
It wouldn't, no. The `wDeprecated` handling here is only for pragma statements, which I believe were only intended to be used with modules, but due to the previous application architecture were (and currently still are) also usable with procedures, like so:
```nim
proc p() =
  var x = 0
  {.deprecated.} # marks `p` as deprecated
```

I think it would make sense to disallow the above usage, hence the todo.

Reply:
Got it, thanks for the explanation.